### ISSUE TITLE
Ability to customize ipv4AllocRange based on node IP

### ISSUE THREAD

**Issue by kekoav** [cilium/cilium#9231]
## Proposal / RFE

**Is your feature request related to a problem?** Yes

The problem we have is the `ipv4AllocRange` selected overlaps with our VPC CIDR range when a node with the last octet in the IP address is the same as `10.x.0.0`. We use a `10.x.0.0` CIDR for our VPCs.

This causes a routing blackhole, and all packets are lost, especially when VXLAN is being used.

We can use the `--ipv4-range` as a workaround, but this range must be specific to a Node, and using this would complicate our Cilium deployment. We can also use the same value if not using VXLAN, but are not sure if this is a good idea either, it feels hacky.

**Describe the solution you'd like**

Create a more configurable way to map Node IP address to the default IPv4 range. Applicable code is here:

https://github.com/cilium/cilium/blob/c6ed206ac554cf803c13e2f7b22269f6fd265432/pkg/defaults/node.go#L28-L34

https://github.com/cilium/cilium/blob/c67da3162445a21fd4af51feef2e4e525fa0d2c8/pkg/node/node_address.go#L80-L100

The translation that happens turns a node IP of the form `x.x.x.y` into `10.y.0.1/16` for the `ipv4AllocRange`.

The desired solution would be a way to customize this mapping to another CIDR. This would allow for dynamic values on each node. Perhaps a new configuration option to specify the CIDR pattern to use. Maybe a Go template? Many variations on this would be acceptable.

**Comment by aanm** [comment]
A workaround for this is to define a pod-cidr that does not colide with the VPC's CIDR. Unfortunately, this is not the case for EKS https://github.com/aws/containers-roadmap/issues/315

**Comment by sleerssen** [comment]
I would like to revisit this request.  With peered VPCs, it's really easy to run into this issue.

**Comment by antoniotamer** [comment]
@kekoav hi - can you tell us if you found a workaround to this issue that worked for you?

**Comment by kekoav** [comment]
@antoniotamer There were a few ways we went about solving this, without fixing in the code (as obvious by this ticket being open still).

1. Hold an IPAM reservation for the problematic IPs. For example, say our CIDR block is `10.80.0.0/16`, we have a problem whenever a node with `x.x.x.80` is created. So we hold a reservation for all subnets that could have a `.80` node in it. We seriously did this for a while to keep the damage from this issue at bay. We created unattached ENIs with IP reservations in AWS to ensure these IPs don't get used by nodes.
2. Move to AWS ENI mode and off VXLAN. This was the ultimate solution for us, which is a step forward in many ways, so we went for it. Obviously, our situation is we are in AWS so that's an option, but it took some time, so we did #1 to stop outages first.
